### PR TITLE
User profile

### PR DESCRIPTION
**Pull request by alexnktn**
Configured user profile with Token and Access levels tabs

### CODE REVIEW

File: grails-app/controllers/UserProfileController.groovy [hunk 1]
```diff
@@ -0,0 +1,100 @@
+import org.transmart.searchapp.AuthUser
+import org.transmartproject.db.log.AccessLogService
+
+class UserProfileController {
+
+	def springSecurityService
```
Comment (burtbeckwith) [inline grails-app/controllers/UserProfileController.groovy]: injected Spring Beans should be typed

File: grails-app/controllers/UserProfileController.groovy [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+	def index() {
+		try{
+			def userDetails
+			def token
+
+			token = auth0Service.jwtToken()
+			userDetails = userService.currentUserInfo()
+
+            model:[
```
Comment (burtbeckwith) [inline grails-app/controllers/UserProfileController.groovy]: no need for `model:`

File: grails-app/controllers/UserProfileController.groovy [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+	def userService
+	def auth0Config
+	AccessLogService accessLogService
+
+	// non-recoverable error
+	String severeMessage = "Unable to update user information. Contact administrator."
+
+	def index() {
+		try{
+			def userDetails
```
Comment (burtbeckwith) [inline grails-app/controllers/UserProfileController.groovy]: Use types

File: grails-app/controllers/UserProfileController.groovy [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+				user: userDetails,
+				token: token?: "Unable to retrieve token.",
+				instanceType: auth0Config.instanceType,
+				instanceName: auth0Config.instanceName,
+				level: userDetails.level
+			]
+
+		}
+		catch (Exception e) {
+			log.error("Caught error in UserProfile plugin.", e);
```
Comment (burtbeckwith) [inline grails-app/controllers/UserProfileController.groovy]: To make the Grails 3 upgrade simpler, use Slf4j instead of the `log` variable injected by Grails. Annotate the class with `@Slf4j('logger')` (the name is overridden to 'logger' to avoid accidentally calling `log`) and use string (not GString) messages with `{}` placeholders, e.g. `logger.error 'Caught error {} in UserProfile plugin', e.message, e`

File: grails-app/controllers/UserProfileController.groovy [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+			redirect(action: 'basic')
+		}
+	}
+
+	def save() {
+
+		try {
+			AuthUser authUser
+			String email     = params.email
```
Comment (burtbeckwith) [inline grails-app/controllers/UserProfileController.groovy]: When the local variable name is the same as the param name, you can declare the variable in the method signature and Grails will set the value and convert to the specified type, e.g. `def save(String email, String firstname, String lastname) {`

File: grails-app/controllers/UserProfileController.groovy [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+			flash.error = severeMessage
+		}
+		redirect(action: 'index')
+	}
+
+	/**
+	 * Basic user profile view if auth0Service not available
+	 */
+	def basic() {
+		AuthUser user = AuthUser.findByUsername(springSecurityService.getPrincipal().username)
```
Comment (burtbeckwith) [inline grails-app/controllers/UserProfileController.groovy]: Should be `AuthUser user = authService.currentAuthUser()`

File: grails-app/controllers/UserProfileController.groovy [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+
+		if (request.post){
+			bindData user, params, [include: [
+					'enabled', 'username', 'userRealName', 'email',
+					'description', 'emailShow', 'authorities', 'changePassword'
+			]]
+			user.name = user.userRealName
+			user.save(flush:true)
+			if (user.hasErrors()) {
+				def errors = ""
```
Comment (burtbeckwith) [inline grails-app/controllers/UserProfileController.groovy]: use `errorStrings()` instead

File: grails-app/services/UserProfileService.groovy [hunk 8]
```diff
@@ -0,0 +1,29 @@
+import grails.transaction.Transactional
+import org.springframework.security.core.authority.SimpleGrantedAuthority
+import org.transmart.searchapp.AuthUser
+import grails.converters.JSON
+
+@Transactional
```
Comment (burtbeckwith) [inline grails-app/services/UserProfileService.groovy]: remove this if it's not really transactional and annotate individual methods that write to the database

File: grails-app/services/UserProfileService.groovy [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+class UserProfileService {
+
+	def springSecurityService
+
+	
+	def getAuthUserLevel() {
+		def roles = springSecurityService.getPrincipal().getAuthorities()
+
+		if (roles.contains(new SimpleGrantedAuthority("ROLE_ADMIN"))){
```
Comment (burtbeckwith) [inline grails-app/services/UserProfileService.groovy]: ick - don't create new SimpleGrantedAuthority instances just to throw them away - extract the authority names and check those

File: grails-app/services/UserProfileService.groovy [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+import org.transmart.searchapp.AuthUser
+import grails.converters.JSON
+
+@Transactional
+class UserProfileService {
+
+	def springSecurityService
+
+	
+	def getAuthUserLevel() {
```
Comment (burtbeckwith) [inline grails-app/services/UserProfileService.groovy]: Why is this needed? Use `authService.currentUserLevel()`

File: grails-app/services/UserProfileService.groovy [hunk 11]
```diff
@@ -0,0 +1,29 @@
+import grails.transaction.Transactional
+import org.springframework.security.core.authority.SimpleGrantedAuthority
+import org.transmart.searchapp.AuthUser
+import grails.converters.JSON
+
+@Transactional
+class UserProfileService {
```
Comment (burtbeckwith) [inline grails-app/services/UserProfileService.groovy]: All new services and src/groovy classes should use `@CompileStatic` except for cases where the code would be significantly more verbose or complicated.

File: grails-app/taglib/transmart/UserProfileTagLib.groovy [hunk 12]
```diff
@@ -0,0 +1,16 @@
+package transmart
+
+class UserProfileTagLib {
+
+    def userProfileService
+
+    def userProfile = { attrs ->
+        if (userProfileService) {
+            out << "<li><a href=\"" + g.createLink(controller:"userProfile", action:"index") + "\">User profile</a></li>"
```
Comment (burtbeckwith) [inline grails-app/taglib/transmart/UserProfileTagLib.groovy]: you should omit `g.` when calling taglib methods in taglibs and controllers. also omit `action:"index"` since it's the default action

File: grails-app/taglib/transmart/UserProfileTagLib.groovy [hunk 13]
```diff
@@ -0,0 +1,16 @@
...
+        }
+        else {
```
Comment (burtbeckwith) [inline grails-app/taglib/transmart/UserProfileTagLib.groovy]: no need to write a blank string

File: grails-app/views/userProfile/basic.gsp [hunk 14]
```diff
@@ -0,0 +1,163 @@
+<!DOCTYPE html>
```
Comment (burtbeckwith) [inline grails-app/views/userProfile/basic.gsp]: omit this in GSPs since it's in the layout

File: grails-app/views/userProfile/basic.gsp [hunk 15]
```diff
@@ -0,0 +1,163 @@
...
+<html lang="en">
+	<head>
+	    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8"/>
```
Comment (burtbeckwith) [inline grails-app/views/userProfile/basic.gsp]: omit this in GSPs since it's in the layout

File: grails-app/views/userProfile/basic.gsp [hunk 16]
```diff
@@ -0,0 +1,163 @@
...
+	    <r:require module="main_mod"/>
+	    <title>User profile</title>
+	    <r:layoutResources/>
+	    
+	    <link rel="stylesheet" type="text/css" href="https://fonts.googleapis.com/css?family=Open+Sans:400,600,700"/>
+	    <link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet">
+		<link rel="stylesheet" href="//maxcdn.bootstrapcdn.com/font-awesome/4.5.0/css/font-awesome.min.css">
+
+        <script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js"></script>
```
Comment (burtbeckwith) [inline grails-app/views/userProfile/basic.gsp]: JS goes at the end of the body

File: grails-app/views/userProfile/basic.gsp [hunk 17]
```diff
@@ -0,0 +1,163 @@
...
+
+        <style>
```
Comment (burtbeckwith) [inline grails-app/views/userProfile/basic.gsp]: move this to a file in web-app/css to allow front-end caching/minification/gzipping/etc.

File: grails-app/views/userProfile/basic.gsp [hunk 18]
```diff
@@ -0,0 +1,163 @@
...
+            }
+        </style>
+
+	</head>
+	
+	
+	
+	<body>
+		<div id="header-div">
+		    <g:render template="/layouts/commonheader" model="['app': '']"/>
```
Comment (burtbeckwith) [inline grails-app/views/userProfile/basic.gsp]: no need for quotes around `app` since Groovy allow you to omit map keys when they're strings

File: grails-app/views/userProfile/basic.gsp [hunk 19]
```diff
@@ -0,0 +1,163 @@
...
+		</div>
+	
+		<div class="body" style="padding-left: 15%">
+		    <h1 style="color: #006dba; font-weight: normal; font-size: 16px; margin: .8em 0 .3em 0;">User Profile</h1>
```
Comment (burtbeckwith) [inline grails-app/views/userProfile/basic.gsp]: avoid using inline styles - use a class, or give the element an id and specify the styles for that id in a .css file

File: grails-app/views/userProfile/basic.gsp [hunk 20]
```diff
@@ -0,0 +1,163 @@
...
+                        </div>
+						<div class="buttons" style="margin-top:20px; margin-bottom:20px">
+							<span class="button">
+								<input class="save" type="submit" value="Save"/>
+							</span>
+						</div>
+					</g:form>
+
+				</section>
+                <section id="content2">
```
Comment (burtbeckwith) [inline grails-app/views/userProfile/basic.gsp]: these should align - indenting should be only tabs (my preference) or only spaces

File: grails-app/views/userProfile/basic.gsp [hunk 21]
```diff
@@ -0,0 +1,163 @@
...
+                    <h3 class="page-header">Access Level</h3>
+                    <div class="row">
+                        <div class="col-lg-12">
+                            <g:if test="${level < 2}">
```
Comment (burtbeckwith) [inline grails-app/views/userProfile/basic.gsp]: don't refer to the numeric level - use the taglibs or the Enum

File: grails-app/views/userProfile/index.gsp [hunk 22]
```diff
@@ -0,0 +1,145 @@
+<!DOCTYPE html>
+<html lang="en">
+	<head>
+	    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8"/>
+	    <r:require module="main_mod"/>
+	    <title>User profile</title>
+	    <r:layoutResources/>
```
Comment (burtbeckwith) [inline grails-app/views/userProfile/index.gsp]: use a layout - no need to copy/paste stuff like this into every GSP. then making changes can happen in one place

**Review by burtbeckwith — CHANGES_REQUESTED** [review]
see inline comments

File: grails-app/controllers/UserProfileController.groovy (hunk 6 above)
Comment (alexnktn) [reply grails-app/controllers/UserProfileController.groovy]: assuming that service isn't available, I think it's fine to leave it like that, there is not much duplicated in here

File: grails-app/controllers/UserProfileController.groovy (hunk 7 above)
Comment (alexnktn) [reply grails-app/controllers/UserProfileController.groovy]: provided custom here in case if UserService isn't there.
